### PR TITLE
Rename the old server before update

### LINKED ISSUE #6602: Failed to download: EPERM: operation not permitted
When I start VSCode it shows me a message that I need to install rust analyzer language server. I click download button and then following error is shown:

```
Failed to download: EPERM: operation not permitted, unlink 'c:\Users\myuser\AppData\Roaming\Code\User\globalStorage\matklad.rust-analyzer\rust-analyzer-windows.exe'
```
I'm using Windows 10, VSCode 1.51.1, extension Rust Analyzer v0.2.384

### PR DESCRIPTION
**Pull request by lnicola**
Kinda' fixes #6602

### CODE REVIEW

File: editors/code/src/net.ts [hunk 1]
```diff
@@ -116,7 +117,42 @@ export async function download(opts: DownloadOpts) {
...
+    }
     await vscode.workspace.fs.rename(tempFilePath, opts.dest, { overwrite: true });
+
+    // Now try to remove any stale server binaries
+    const serverDir = vscode.Uri.file(rawDest.dir);
+    try {
+        const entries = await vscode.workspace.fs.readDirectory(serverDir);
+        for (const [entry, _] of entries) {
+            try {
+                const parsedPath = path.parse(entry);
```
Comment (lnicola) [inline editors/code/src/net.ts:139]: Not sure what `entry` is here.

**Comment by wxb1ank** [comment]
Sorry if I'm missing something, but how is temporarily renaming the server to 'stale' different than replacing it with the newly downloaded one? I haven't tested, but if one or more server instances are running, won't this always cause EPERM?

**Comment by lnicola** [comment]
It seems that you actually _can_ rename a running executable on Windows. But yeah, this is untested otherwise.

**Comment by wxb1ank** [comment]
I feel like we wouldn't have run into the original permission error if that were the case. IMO, the simplest thing to do here is to check specifically for EPERM and log something ala "Failed to overwrite the old server executable. Are there instances currently running?". Needing to update RA while it's running seems like a niche case to me.

**Comment by lnicola** [comment]
Yeah, I'd agree, but this seems to bother a lot of people. So while it's not a full solution, at least it might take us 90% of the way there.

**Comment by wxb1ank** [comment]
Well, the code LGTM otherwise. The first Google result I found on this topic implied this couldn't be done, but [this](https://superuser.com/questions/488127/why-can-i-rename-a-running-executable-but-not-delete-it) suggests otherwise? I'll try and test this later today or tomorrow.

File: editors/code/src/net.ts [hunk 2]
```diff
@@ -116,7 +117,41 @@ export async function download(opts: DownloadOpts) {
...
+    // NB: this can lead to issues if a running Code instance tries to restart the server.
+    try {
+        await vscode.workspace.fs.rename(opts.dest, oldServerPath, { overwrite: true });
+        log.info(`Renamed old server binary ${opts.dest.fsPath} to ${oldServerPath.fsPath}`);
+    } catch (err) {
+        const fsErr = err as vscode.FileSystemError;
+        if (!fsErr || fsErr.code !== "FileNotFound") {
+            log.error(`Cannot rename running instance: ${err}`);
+        }
+    }
```
Comment (wxb1ank) [inline editors/code/src/net.ts:130]: What does `!fsErr` do here?

Comment (lnicola) [reply editors/code/src/net.ts:130]: I'm checking for null, in case it's a different exception.

Comment (wxb1ank) [reply editors/code/src/net.ts:130]: The reminder in [this section](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) of the TS docs seems to suggest that `fsErr` can never be null.

Comment (lnicola) [reply editors/code/src/net.ts:130]: Oops, thanks. I think Error has a name field, maybe that's what it's supposed to be used for?

**Comment by wxb1ank** [comment]
I tested with two VS Code windows open. One was running RA in the background while the other attempted to update.

```
DEBUG [7/12/2021, 5:57:01 PM]: Downloading file of 254131 bytes size from https://github.com/rust-analyzer/rust-analyzer/releases/download/nightly/rust-analyzer.vsix to c:\Users\wxblank\AppData\Roaming\Code\User\globalStorage\matklad.rust-analyzer\rust-analyzer169d3161c8
ERROR [7/12/2021, 5:57:02 PM]: Cannot rename running instance: EntryNotFound (FileSystemError) (FileSystemError): Error: ENOENT: no such file or directory, rename 'c:\Users\wxblank\AppData\Roaming\Code\User\globalStorage\matklad.rust-analyzer\rust-analyzer.vsix' -> 'c:\Users\wxblank\AppData\Roaming\Code\User\globalStorage\matklad.rust-analyzer\rust-analyzer-stale-169d3161c8.vsix'
ERROR [7/12/2021, 5:57:02 PM]: Bootstrap error [Error: ENOENT: no such file or directory, open 'c:\Users\wxblank\AppData\Roaming\Code\User\globalStorage\matklad.rust-analyzer\rust-analyzer.vsix',    at h (file:///C:/Program Files/Microsoft VS Code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:54:176162),    at file:///C:/Program Files/Microsoft VS Code/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:54:177561,    at C:\Program Files\Microsoft VS Code\resources\app\node_modules.asar\yauzl\index.js:34:21,    at FSReqCallback.oncomplete (fs.js:171:23)]
```

https://github.com/rust-analyzer/rust-analyzer/blob/5280b6d84c4fc727d055b02239d631963264cb2b/editors/code/src/net.ts#L127
It seems this doesn't work. VS Code docs strike again! `EntryNotFound` was thrown instead of the expected `FileNotFound`. We *could* just hardcode the first one instead, but I think a more resilient solution is as follows:

1. Catch and ignore this particular error.
2. Catch and handle the error from replacing the old binary with the new, downloaded one. Currently, [we don't catch this one at all](https://github.com/rust-analyzer/rust-analyzer/blob/5280b6d84c4fc727d055b02239d631963264cb2b/editors/code/src/net.ts#L131), and I think it's the reason I logged a bootstrap error. However, I'm a little confused why I get an error at all.

**Comment by wxb1ank** [comment]
With the following:

```ts
// Try to rename a running server to avoid EPERM on Windows
// NB: this can lead to issues if a running Code instance tries to restart the server.
try {
    await vscode.workspace.fs.rename(opts.dest, oldServerPath, { overwrite: true });
    log.info(`Renamed old server binary ${opts.dest.fsPath} to ${oldServerPath.fsPath}`);
} catch (err) {

}

try {
    await vscode.workspace.fs.rename(tempFilePath, opts.dest, { overwrite: true });
} catch (err) {
    log.error(`Failed to replace the old binary with the new one: ${err}`);
}
```

I no longer receive the previous error, and the update seems to go smoothly (with the old RA instance still active).

**Comment by lnicola** [comment]
I ended up checking for both, to avoid showing an error when there's no existing binary, but still complain when something happens. Of course `vscode.workspace.fs.stat` also throws on missing files.

Also, thanks a lot for giving this a try!

**Comment by lnicola** [comment]
I think I want to merge this today so we can test it a little before next week (unless someone r-'es it, of course).

**Pull request merged**